src/main/java/org/opensearch/neuralsearch/query/SparseEncodingQueryBuilder.java

```diff
@@ -0,0 +1,306 @@
+                        parser.getTokenLocation(),
+                        "[" + NAME + "] query does not support [" + currentFieldName + "]"
+                    );
+                }
+            } else if (QUERY_TOKENS_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                sparseEncodingQueryBuilder.queryTokens(parser.map(HashMap::new, XContentParser::floatValue));
+            } else {
+                throw new ParsingException(
+                    parser.getTokenLocation(),
+                    "[" + NAME + "] unknown token [" + token + "] after [" + currentFieldName + "]"
+                );
+            }
+        }
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        // If the user has specified query_tokens field, then we don't need to inference the sentence,
+        // just re-rewrite to self. Otherwise, we need to inference the sentence to get the queryTokens. Then the
+        // logic is similar to NeuralQueryBuilder
+        if (null != queryTokens) {
+            return this;
+        }
+        if (null != queryTokensSupplier) {
+            return queryTokensSupplier.get() == null
+                ? this
+                : new SparseEncodingQueryBuilder().fieldName(fieldName)
+                    .queryTokens(queryTokensSupplier.get())
+                    .queryText(queryText)
+                    .modelId(modelId);
+        }
+
+        validateForRewrite(queryText, modelId);
+        SetOnce<Map<String, Float>> queryTokensSetOnce = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
+                modelId(),
+                List.of(queryText),
+                ActionListener.wrap(mapResultList -> {
+                    queryTokensSetOnce.set(TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0));
+                    actionListener.onResponse(null);
+                }, actionListener::onFailure)
+            ))
+        );
+        return new SparseEncodingQueryBuilder().fieldName(fieldName)
+            .queryText(queryText)
+            .modelId(modelId)
+            .queryTokensSupplier(queryTokensSetOnce::get);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        final MappedFieldType ft = context.fieldMapper(fieldName);
+        validateFieldType(ft);
+        validateQueryTokens(queryTokens);
+
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        for (Map.Entry<String, Float> entry : queryTokens.entrySet()) {
+            builder.add(FeatureField.newLinearQuery(fieldName, entry.getKey(), entry.getValue()), BooleanClause.Occur.SHOULD);
+        }
+        return builder.build();
+    }
+
+    private static void validateForRewrite(String queryText, String modelId) {
+        if (null == queryText || null == modelId) {
```

Comment:
Please use StringUtils.isBlank to avoid unnecessary invocations, empty queryText doesn't have concrete meaning from business perspective.

Reply:
Sure